File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/commands/RequestPayment.java
package it.giannibombelli.wsc2026.payment.application.commands;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Command;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;

public record RequestPayment(PaymentId aggregateId, ClientReference clientReference, Money amount, Timestamp requestedAt)
    implements Command<PaymentId> {
    public RequestPayment {
        Require.requireArgument(aggregateId, "paymentId");
        Require.requireArgument(clientReference, "clientReference");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(requestedAt, "requestedAt");
    }
}

File: typescript/src/payment/application/commands/requestPayment.ts
// Command per richiedere un nuovo pagamento.

import type { Command } from '@/common/application/command.js';
import type { ClientReference } from '@/common/domain/primitive/clientReference.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { Timestamp } from '@/common/domain/primitive/timestamp.js';
import type { PaymentId } from '../../domain/payment/paymentId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type RequestPayment = Command<'RequestPayment', PaymentId> & {
  readonly clientReference: ClientReference;
  readonly amount: Money;
  readonly requestedAt: Timestamp;
};

export function requestPayment(
  aggregateId: PaymentId,
  clientReference: ClientReference,
  amount: Money,
  requestedAt: Timestamp
): RequestPayment {
  requireArgument(aggregateId, 'paymentId');
  requireArgument(clientReference, 'clientReference');
  requireArgument(amount, 'amount');
  requireArgument(requestedAt, 'requestedAt');
  return { kind: 'RequestPayment', aggregateId, clientReference, amount, requestedAt };
}
